wolfcrypt/src/random.c

```diff
@@ -1403,20 +1411,23 @@ int wc_FreeNetRandom(void)
         return (ok) ? 0 : -1;
     }
 
-#else /* USE_WINDOWS_API */
+#else /* USE_INTEL_INTRINSICS */
     /* The compiler Visual Studio uses does not allow inline assembly.
      * It does allow for Intel intrinsic functions. */
 
     /* return 0 on success */
+# ifdef __GNUC__
+    __attribute__((target("rdseed")))
+# endif
     static WC_INLINE int IntelRDseed64(word64* seed)
     {
         int ok;
 
-        ok = _rdseed64_step(seed);
+        ok = _rdseed64_step((unsigned long long*) seed);
```

Comment:
Note: Seeing this test error on our AMD Jenkins rig:

```
./configure --enable-intelasm
wolfcrypt/src/random.c: In function ‘IntelRDseed64’:
wolfcrypt/src/random.c:1426:14: error: implicit declaration of function ‘_rdseed64_step’; did you mean ‘_rdrand64_step’? [-Werror=implicit-function-declaration]
         ok = _rdseed64_step((unsigned long long*) seed);
              ^~~~~~~~~~~~~~
              _rdrand64_step
wolfcrypt/src/random.c:1426:14: error: nested extern declaration of ‘_rdseed64_step’ [-Werror=nested-externs]
cc1: all warnings being treated as errors
make[1]: *** [wolfcrypt/src/src_libwolfssl_la-random.lo] Error 1
```

@dr-m , any ideas?

Thanks,
David Garske, wolfSSL

Reply:
Of particular note: this error occurred on a PC with an AMD chipset. AMD ALSO supports intel intrinsics so we test both AMD and Intel chipsets with intelasm:

CPU info:

```
vendor_id	: AuthenticAMD
cpu family	: 23
model		: 8
model name	: AMD Ryzen 7 2700X Eight-Core Processor
```

Reply:
@dgarske, what is the compiler version, and what are the command line parameters passed to the compiler? I had made some experiments on https://godbolt.org with different compiler versions, and it seemed to me that the `__attribute__((target))` as well as the Intel intrinsic header were OK starting with GCC 5.1 (there was no GCC 5.0 available).

So far, I did not run the `./configure` script; I only built WolfSSL as part of the MariaDB server, using our CMake scripts. But I am happy to troubleshoot if you can give more details.

For the record, most builders on our old CI system at https://buildbot.askmonty.org are virtual machines, using `--cpu=qemu64`, which seems to make `cpuid` claim that no extensions are supported on top of the original AMD64 ISA. But, as far as I remember, our build scripts are not passing any `-march` or `-mtune` parameters, so this should not make any difference to compilation, like it might if something like `-march=native` were used.

Reply:
On my system, with GCC 10.2.0, I tried the following:
```sh
./autogen.sh
./configure --enable-intelasm --enable-intelrand
make
touch wolfcrypt/src/random.c
make AM_DEFAULT_VERBOSITY=1
```
The last command revealed the compiler invocation:
```
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -DBUILDING_WOLFSSL -DBUILDING_WOLFSSL -DWOLFSSL_TLS13 -DHAVE_TLS_EXTENSIONS -DHAVE_SUPPORTED_CURVES -DHAVE_FFDHE_2048 -D_POSIX_THREADS -fvisibility=hidden -DHAVE_THREAD_LS -DNDEBUG -pthread -DTFM_TIMING_RESISTANT -DECC_TIMING_RESISTANT -DWC_RSA_BLINDING -DHAVE_AESGCM -DWOLFSSL_AESNI -maes -msse4 -mpclmul -DHAVE_INTEL_RDSEED -DUSE_INTEL_SPEEDUP -DHAVE_INTEL_RDRAND -DWOLFSSL_SHA512 -DWOLFSSL_SHA384 -DHAVE_HKDF -DNO_DSA -DHAVE_ECC -DTFM_ECC256 -DECC_SHAMIR -DWC_RSA_PSS -DWOLFSSL_BASE64_ENCODE -DNO_RC4 -DNO_HC128 -DNO_RABBIT -DWOLFSSL_SHA224 -DWOLFSSL_SHA3 -DWOLFSSL_SHAKE256 -DHAVE_POLY1305 -DHAVE_ONE_TIME_AUTH -DHAVE_CHACHA -DHAVE_HASHDRBG -DHAVE_TLS_EXTENSIONS -DHAVE_SUPPORTED_CURVES -DHAVE_EXTENDED_MASTER -DNO_RC4 -DHAVE_ENCRYPT_THEN_MAC -DNO_PSK -DNO_MD4 -DNO_PWDBASED -DUSE_FAST_MATH -DWOLFSSL_X86_64_BUILD -DWC_NO_ASYNC_THREADING -DHAVE_DH_DEFAULT_PARAMS -DNO_DES3 -Wall -Wno-unused -O2 -DHAVE___UINT128_T=1 -Werror -Wno-pragmas -Wall -Wno-strict-aliasing -Wextra -Wunknown-pragmas --param=ssp-buffer-size=1 -Waddress -Warray-bounds -Wbad-function-cast -Wchar-subscripts -Wcomment -Wfloat-equal -Wformat-security -Wformat=2 -Wmaybe-uninitialized -Wmissing-field-initializers -Wmissing-noreturn -Wmissing-prototypes -Wnested-externs -Wnormalized=id -Woverride-init -Wpointer-arith -Wpointer-sign -Wredundant-decls -Wshadow -Wsign-compare -Wstrict-overflow=1 -Wswitch-enum -Wundef -Wunused -Wunused-result -Wunused-variable -Wwrite-strings -fwrapv -MT wolfcrypt/src/src_libwolfssl_la-random.lo -MD -MP -MF wolfcrypt/src/.deps/src_libwolfssl_la-random.Tpo -c wolfcrypt/src/random.c  -fPIC -DPIC -o wolfcrypt/src/.libs/src_libwolfssl_la-random.o
```
By invoking the command
```sh
objdump -d wolfcrypt/src/.libs/src_libwolfssl_la-random.o
```
I see that the instructions are being generated, at the start of the `.text` section:
```
wolfcrypt/src/.libs/src_libwolfssl_la-random.o:     file format elf64-x86-64


Disassembly of section .text:

0000000000000000 <IntelRDrand64>:
       0:       48 0f c7 f0             rdrand %rax
       4:       ba 01 00 00 00          mov    $0x1,%edx
       9:       48 89 07                mov    %rax,(%rdi)
       c:       0f 42 c2                cmovb  %edx,%eax
       f:       85 c0                   test   %eax,%eax
      11:       0f 94 c0                sete   %al
      14:       0f b6 c0                movzbl %al,%eax
      17:       f7 d8                   neg    %eax
      19:       c3                      retq   
      1a:       66 0f 1f 44 00 00       nopw   0x0(%rax,%rax,1)

0000000000000020 <IntelRDseed64>:
      20:       48 0f c7 f8             rdseed %rax
```
Can you please give more details on the compilation failure in your environment?

Reply:
@dr-m:

That test machine builds with both:
gcc version 8.1.0 (GCC)
clang version 6.0.0-1ubuntu2

Auto-tools versions are:
autoconf (GNU Autoconf) 2.69
automake (GNU automake) 1.15.1
libtoolize (GNU libtool) 2.4.6

The configuration that failed was simply:

```
./configure --enable-intelasm
```

- KH

Reply:
I was able to reproduce this in a Debian Buster `chroot` environment, using `clang-6.0`. Apparently, the correct header file to include is `x86intrin.h` instead of `immintrin.h`, and the preprocessor symbol `__RDSEED__` must be enabled in order to define the Intel intrinsic function. It looks like with a newer set of header files, including `immintrin.h` is sufficient.

With some fixes, I was able to compile with
```sh
CC=clang-6.0 ./configure --enable-intelasm --enable-intelrand
```
and both instructions are present in `wolfcrypt/src/.libs/src_libwolfssl_la-random.o`. If I omit the `--enable-intelrand`, the `rdrand` instruction will not be emitted.

However, GCC 8.3.0 triggered a link error:
```
/usr/bin/ld: src/.libs/libwolfssl.so: undefined reference to `__builtin_ia32_rdseed_di_step'
```
Finally, the following worked for me:
```diff
diff --git a/wolfcrypt/src/random.c b/wolfcrypt/src/random.c
index 0f7b60195..49d407ebe 100644
--- a/wolfcrypt/src/random.c
+++ b/wolfcrypt/src/random.c
@@ -193,6 +193,23 @@ int wc_RNG_GenerateByte(WC_RNG* rng, byte* b)
 
 #ifdef USE_INTEL_INTRINSICS
     #include <immintrin.h>
+    /* clang-6.0 or gcc-8 would not define _rdseed64_step() in immintrin.h */
+    #ifndef HAVE_INTEL_RDSEED
+    #elif defined __clang__ && __clang_major__ > 6
+    #elif !defined __GNUC__
+    #elif __GNUC__ > 8
+    #else
+        #ifndef __clang__
+            #pragma GCC push_options
+            #pragma GCC target("rdseed")
+        #else
+            #define __RDSEED__
+        #endif
+        #include <x86intrin.h>
+        #ifndef __clang__
+            #pragma GCC pop_options
+        #endif
+    #endif
 #endif /* USE_WINDOWS_API */
 #endif
 
```
I will validate this solution on https://godbolt.org on GCC 5 through 10 and clang 4 through 10, to find out whether any further adjustments are needed, and I will update my patch accordingly.

Reply:
It looks like the `_rdrand64_step()` was introduced in `clang-3.8`. The `#include <immintrin.h>` defines both functions starting with `clang-7` or `gcc-9`. I tested the following:
```c
//#define HAVE_INTEL_RDSEED
    #include <immintrin.h>
    #ifndef HAVE_INTEL_RDSEED
    #elif defined __clang__ && __clang_major__ > 6
    #elif !defined __GNUC__
    #elif __GNUC__ > 8
    #else
        #ifndef __clang__
            #pragma GCC push_options
            #pragma GCC target("rdseed")
        #else
            #define __RDSEED__
        #endif
        #include <x86intrin.h>
        #ifndef __clang__
            #pragma GCC pop_options
        #endif
    #endif

__attribute__((target("rdrnd")))
int IntelRDrand64(unsigned long long *rnd) { return _rdrand64_step(rnd); }
__attribute__((target("rdseed")))
int IntelRDseed(unsigned long long *rnd) { return _rdseed64_step(rnd); }
```
Uncomment the first line to make the code compile with anything between Clang 3.8 and 7, or between GCC 5 and 8. Apparently, the function for `rdrand` was always available via `immintrin.h` on these compilers; only `rdseed` needs the extra hoops.